### ISSUE TITLE
Support more types for `ScriptModule`

### ISSUE THREAD

**Issue by kaiidams** [dotnet/TorchSharp#771]
PyTorch mobile's Java binding exports a wrapper of `IValue` (https://pytorch.org/javadoc/1.9.0/)

The Python script below makes a TorchScript file that uses more type for testing.

```python
import torch
from torch import nn

# Used types:
#   bool
#   bool_list
#   dict_long_key
#   dict_string_key
#   double
#   double_list
#   list
#   long
#   long_list
#   null
#   string
#   tensor
#   tensor_list
#   tuple


class MyModule(nn.Module):
    @torch.jit.export
    def forward(
        self,
        x: torch.Tensor,
        d: float,
        n: int,
        s: str,
    ):
        return (
            True,
            {
                1: [False, True],
                2: 1.2,
                n: [2.3, d],
            },
            {
                "abc": [123, torch.arange(n), None],
                s: [456, 789]
            },
            [x, torch.arange(n)]
        )


module = torch.jit.script(MyModule())
print(module.forward.code)
module.save("ivalue_test.pt")
module = torch.jit.load("ivalue_test.pt")
print(module.code)
```

**Comment by NiklasGustafsson** [comment]
@kaiidams: how essential is this compared with the other ScriptModule issues (which I already have fixes for)? I think this could take a bit of work, and would hold up making another release with the new forward(), new losses, and other ScriptModule work.

**Comment by kaiidams** [comment]
@NiklasGustafsson 

Tuple responses with `Tensor`s and `None` for optional values are most common, but supporting all types are not that important. I don't think you have to hold the next release for this issue.

`ScriptModule.forward()` can be implemented like

```
IValue forward(IValue[] args, IDictionary<string, IValue> kwargs)
```

which is closer to LibTorch C++/Java. Or

```
object forward(object[] args, IDictionary<string, object> kwargs)
```

which returns C# native values like `IDictionary` or `IList`.

**Comment by NiklasGustafsson** [comment]
@kaiidams -- limited support for other types (scalars) has been released in 0.98.0, along with all the forward, loss function, script module methods, jit.compile, etc. work.

**Comment by kaiidams** [comment]
@NiklasGustafsson Should be there another github issue for tracking other types like a dictionary?
